src/gt4py/storage/allocators.py

```diff
@@ -0,0 +1,421 @@
+
+        # Compute number of items inside an aligned block
+        item_size = dtype.byte_size
+        if math.gcd(byte_alignment, item_size) not in (byte_alignment, item_size):
+            raise ValueError(
+                f"Incompatible 'byte_alignment' ({byte_alignment}) and 'dtype' size ({item_size})"
+            )
+        items_per_aligned_block = (byte_alignment // item_size) or 1
+
+        # Compute the padding required in the contiguous dimension to get aligned blocks
+        dims_layout = [layout_map.index(i) for i in range(len(shape))]
+        padded_shape_lst = list(shape)
+        padded_shape_lst[dims_layout[-1]] = (
+            math.ceil(shape[dims_layout[-1]] / items_per_aligned_block) * items_per_aligned_block
+        )
+        padded_shape = tuple(padded_shape_lst)
+        assert core_defs.is_valid_tensor_shape(padded_shape)
+        total_length = item_size * functools.reduce(operator.mul, padded_shape) + (
+            byte_alignment - 1
+        )  # the worst case for misalignment is `byte_alignment - 1`
+
+        # Compute strides
+        strides_lst = [item_size] * len(shape)
+        accumulator = item_size
+        for i in range(len(shape) - 2, -1, -1):
+            accumulator = strides_lst[dims_layout[i]] = (
+                accumulator * padded_shape[dims_layout[i + 1]]
+            )
+        strides = tuple(strides_lst)
+
+        # Allocate total size
+        buffer = self.raw_alloc(total_length, device)
+        memory_address = self.array_ns.byte_bounds(buffer)[0]
+
+        # Compute final byte offset to align the requested buffer index
+        aligned_index = tuple(aligned_index or ([0] * len(shape)))
+        aligned_index_offset = (
+            items_per_aligned_block
+            * (int(math.ceil(aligned_index[dims_layout[-1]] / items_per_aligned_block)))
+            - aligned_index[dims_layout[-1]]
+        ) * item_size
+
+        allocation_mismatch_offset = (
+            byte_alignment - memory_address % byte_alignment
+        ) % byte_alignment
+        byte_offset = (aligned_index_offset + allocation_mismatch_offset) % byte_alignment
+
+        # Create shaped view from buffer
+        ndarray = self.tensorize(
+            buffer, dtype, shape, padded_shape, item_size, strides, byte_offset
+        )
+
+        if device.device_type == core_defs.DeviceType.ROCM:
+            # until we can rely on dlpack
+            ndarray.__hip_array_interface__ = {  # type: ignore[attr-defined]
+                "shape": buffer.ndarray.shape,  # type: ignore[union-attr]
+                "typestr": buffer.ndarray.dtype.descr[0][1],  # type: ignore[union-attr]
+                "descr": buffer.ndarray.dtype.descr,  # type: ignore[union-attr]
+                "stream": 1,
+                "version": 3,
+                "strides": buffer.ndarray.strides,  # type: ignore[union-attr]
+                "data": (buffer.ndarray.data.ptr, False),  # type: ignore[union-attr]
+            }
```

Comment:
I think it should be:
```suggestion
            ndarray.__hip_array_interface__ = {  # type: ignore[attr-defined]
                "shape": ndarray.shape,  # type: ignore[union-attr]
                "typestr": ndarray.dtype.descr[0][1],  # type: ignore[union-attr]
                "descr": ndarray.dtype.descr,  # type: ignore[union-attr]
                "stream": 1,
                "version": 3,
                "strides": ndarray.strides,  # type: ignore[union-attr]
                "data": (ndarray.data.ptr, False),  # type: ignore[union-attr]
            }
```